src/uct/tcp/tcp_net.c

```diff
@@ -184,38 +214,67 @@ ucs_status_t uct_tcp_netif_is_default(const char *if_name, int *result_p)
     return UCS_OK;
 }
 
-static ucs_status_t uct_tcp_do_io(int fd, void *data, size_t *length_p,
-                                  uct_tcp_io_func_t io_func, const char *name)
+int uct_tcp_sockaddr_cmp(const struct sockaddr *sa1,
+                         const struct sockaddr *sa2)
 {
-    ssize_t ret;
-
-    ucs_assert(*length_p > 0);
-    ret = io_func(fd, data, *length_p, MSG_NOSIGNAL);
-    if (ret == 0) {
-        ucs_trace("fd %d is closed", fd);
-        return UCS_ERR_CANCELED; /* Connection closed */
-    } else if (ret < 0) {
-        if ((errno == EINTR) || (errno == EAGAIN)) {
-            *length_p = 0;
-            return UCS_OK;
-        } else {
-            ucs_error("%s(fd=%d data=%p length=%zu) failed: %m",
-                      name, fd, data, *length_p);
-            return UCS_ERR_IO_ERROR;
-        }
-    } else {
-        *length_p = ret;
-        return UCS_OK;
+    int cmp;
+
+    if (sa1->sa_family != sa2->sa_family) {
+        return memcmp(&sa1->sa_family, &sa2->sa_family,
+                      sizeof(sa1->sa_family));
+    }
+
+    switch (sa1->sa_family) {
+    case AF_INET:
+        cmp = memcmp(&UCS_SIN_ADDR(sa1), &UCS_SIN_ADDR(sa2),
+                     sizeof(UCS_SIN_ADDR(sa1)));
+        return cmp ? cmp : memcmp(&UCS_SIN_PORT(sa1),
+                                  &UCS_SIN_PORT(sa2),
+                                  sizeof(UCS_SIN_PORT(sa1)));
+    case AF_INET6:
+        cmp = memcmp(&UCS_SIN6_ADDR(sa1), &UCS_SIN6_ADDR(sa2),
+                     sizeof(UCS_SIN6_ADDR(sa1)));
+        return cmp ? cmp : memcmp(&UCS_SIN6_PORT(sa1),
+                                  &UCS_SIN6_PORT(sa2),
+                                  sizeof(UCS_SIN6_PORT(sa1)));
+    default:
+        ucs_error("unknown address family: %d", sa1->sa_family);
+        return 0;
```

Comment:
does this mean success?

Reply:
hmm, we should return something better here... but <0, 0, >0 are already used.
probably we can return only -1, 0, 1 - because only them are checked in TCP CM

Reply:
fixed. Now it returns `signum(memcmp(...))`